keps/sig-node/3675-resource-plugin-manager/README.md

```diff
@@ -0,0 +1,1229 @@
+plugins.
+
+##### Example 1: Attributed-based resource specification
+
+One realization of such an API can be achieved by an attribute-based 
+mechanism for compute resources that resembles the Dynamic Resource 
+Allocation (DRA) claim mechanism. The attribute-based compute resource 
+request model is a step towards improved semantic workload portability 
+that can be tried in a progressive manner with CCI. 
+
+In the following, there is an example of a novel approach that could be 
+fulfilled with the CCI architecture. This example presents a set of
+per-core attributes that enable precision beyond a policy-only based approach: 
+
+    # A resource request which consists of:
+    # 2 exclusive cores (no other processes running on them)
+    # 	* the cores shall be siblings
+    # 	* they shall run at 2.1 ghz
+    # 	* application assigned has the highest priority on the core
+    # 	* there are no IRQs on the core
+    # 6 exclusive cores(no other processes running on them)
+    # 4 shared cores(other processes can run on them) 
+    apiVersion: v1
+    kind: ConfigMap
+    metadata:
+      name: example-compute-claim-parameters
+      namespace: default
+    data:
+      cores: |
+    2
+    6
+    4
+      compute_attributes: |
+        exclusive, smt-sibling-required, 2.1Ghz, 1.0 priority, no-irq
+        exclusive
+        shared
+    apiVersion: resource.k8s.io/v1alpha1
+    kind: ResourceClaimTemplate
+    metadata:
+      name: example-compute-claim-parameters-template
+      namespace: default
+    spec:
+      metadata:
+        labels:
+          app: inline-resource
+      spec:
+        resourceClassName: cpuressources
+        parametersRef:
+          kind: ConfigMap
+          name: example-compute-claim-parameters
+
+##### Example 2.: Policy-based resource specification
+
+Another option how a resource allocation configuration for Kubernetes
+Pods could be achieved is through a policy mechanism. In this case the
+policy will be applied to all Pods to be processed. The configuration
+of the policy can happen directly in a CCI Compute Resource Driver Plugin
+as input argument in the yaml deployment. This methodology can be used to
+realize methodologies similar to the static CPU manager policy or other policies.
+
+##### Resource Manager Architecture
+
+The extension consists of an optional CCI driver inside Pod spec which
+indicates if a CCI driver exists, then it shall process compute resources. 
+The CCI plugin allocations can coexist with the existing CPU manager and 
+memory manager allocations through the introduction of a new policy inside 
+the CPU manager: cci-policy. The policy addition will allow the non-disruptive
+handling of all existing Pods without attached CCI drivers by the existing
+CPU manager and the triggering of Pod allocation done via the driver.
+The policy also takes care for fallback options, proper cluster start and 
+plugin failure handling.
```

Comment:
This section seems to imply that traditional CPU/RAM management is meant to co-exist with management through CCI.

I am worried that this will lead to race conditions that defeat one of the promises made by DRA ("pods don't get scheduled unless resources are available on the node").

Suppose a node is unused and two pods need to be scheduled, one with CCI and one without. Both will completely fill the node. Depending on how the race ends, the ResourceClaim for the CCI pod might get marked as allocated in parallel to scheduling the traditional pod to the node. Then both got scheduled, but only one can run.

Reply:
Our objective is to synchronize the states of cpu manager and CCI Drivers to avoid race conditions like that. The implementation have to make sure that such race cases are prohibited either through sequential code or through proper synchronization concepts.

Reply:
The traditional assignment of CPU and RAM to pods by kube-scheduler is done without synchronization. That's why it is fast and can schedule pods rapidly.

I don't think it is feasible to add reliable synchronization without affecting "normal" pod scheduling.

Reply:
Just to be clear: synchronizing on the node inside kubelet may be feasible. It's just not enough because it doesn't include kube-scheduler.

Reply:
we focus on on-node solution after scheduling decision, admission of pods (CCI or std) have to be synchronous